sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2133,6 +2096,28 @@ void ProgramManager::removeImages(sycl_device_binaries DeviceBinary) {
     // Acquire lock to modify maps for kernel bundles
     std::lock_guard<std::mutex> KernelIDsGuard(m_KernelIDsMutex);
 
+    {
+      // Clean up Bfloat16 device library image, unregister exported symbols
+      // and remove the device image only when RefCount is 0.
+      auto Bfloat16DeviceLibProp = Img->getDeviceLibMetadata();
+      if (Bfloat16DeviceLibProp.isAvailable()) {
+        uint32_t IsNative =
+            DeviceBinaryProperty(*(Bfloat16DeviceLibProp.begin())).asUint32();
+        if (m_Bfloat16DeviceLibRefCount[IsNative] != 0)
+          m_Bfloat16DeviceLibRefCount[IsNative] -= 1;
+        if (!m_Bfloat16DeviceLibRefCount[IsNative]) {
+          for (const sycl_device_binary_property &ESProp :
+               Img->getExportedSymbols()) {
+            m_ExportedSymbolImages.erase(ESProp->Name);
+          }
+
+          m_DeviceImages.erase(DevImgIt);
```

Comment:
Not erasing the device image here when the refcount > 0 won't keep the underlying device binaries alive. Consider the following situation:

- bundle A loaded, contributes bfloat dev lib into runtime -> refcount = 1
- bundle B loaded, uses bfloat dev lib -> refcount = 2
- bundle A is freed -> refcount = 1
- bundle C loaded, uses bfloat dev lib -> refcount = 2, and crash when PM tries to link the kernels because `m_ExportedSymbols` points to image coming from bundle A, which has been destroyed